File: dsa-problems/GraphsBasics/solutions/javascript/GRB-005-topo-sort-kahn.js
const readline = require("readline");

class Solution {
  topoSort(n, adj) {
    const indegree = new Int32Array(n).fill(0);
    for (let u = 0; u < n; u++) {
      for (const v of adj[u]) {
        indegree[v]++;
      }
    }

    const queue = [];
    for (let i = 0; i < n; i++) {
      if (indegree[i] === 0) {
        queue.push(i);
      }
    }

    const result = [];
    let head = 0; // Pointer for O(1) dequeue
    while (head < queue.length) {
      const u = queue[head++];
      result.push(u);

      for (const v of adj[u]) {
        indegree[v]--;
        if (indegree[v] === 0) {
          queue.push(v);
        }
      }
    }

    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const m = parseInt(data[idx++], 10);
  
  const adj = Array.from({ length: n }, () => []);
  for (let i = 0; i < m; i++) {
    const u = parseInt(data[idx++], 10);
    const v = parseInt(data[idx++], 10);
    adj[u].push(v);
  }

  const solution = new Solution();
  const order = solution.topoSort(n, adj);
  console.log(order.join(" "));
});

File: dsa-problems/GraphsBasics/solutions/python/GRB-005-topo-sort-kahn.py
import sys
from collections import deque

def topo_sort(n: int, adj: list[list[int]]) -> list[int]:
    indegree = [0] * n
    for u in range(n):
        for v in adj[u]:
            indegree[v] += 1
            
    queue = deque([i for i in range(n) if indegree[i] == 0])
    result = []
    
    while queue:
        u = queue.popleft()
        result.append(u)
        
        for v in adj[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                queue.append(v)
                
    return result

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        m = int(next(iterator))
        adj = [[] for _ in range(n)]
        for _ in range(m):
            u = int(next(iterator))
            v = int(next(iterator))
            adj[u].append(v)
            
        order = topo_sort(n, adj)
        print(" ".join(map(str, order)))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
